### ISSUE TITLE
Building from source is a hodgepodge compared to standard Go apps

### ISSUE THREAD

**Issue by braderhart** [algorand/go-algorand#2131]
## Summary

This all started when I just wanted to get started using the CLI tools, and the instructions on the website suggested that I run a script on my computer. After following the script, it was just too convoluted for me to reliably trust that it would even be a good idea. I have built multiple Go applications from source, including those that use C libraries. 

When building go-algorand from source, it is currently expected that people run a Makefile that is chained to multiple Bash scripts within Bash scripts. This is contrary to standard Go packages that should be as simple as running `go install ./cmd/goal`. I personally think it has just been over-engineered. For releases, there are examples within the [Hugo repo](https://github.com/gohugoio/hugo/blob/master/goreleaser.yml) on ways to simplify.

Documentation often is lot more useful than putting everything into a script. Expecting users wanting to build from source to run a bunch of foreign code that is a pain to follow in order eventually run a simple `go build/install` should be concerning. As a developer approaching any project like this I have to ask myself why are things so obfuscated from standard practices, and is it really necessary?

There is also a concern about how you are implementing `libsodium`. Currently, the patched version is committed directly to the repo. This makes it difficult to see the changes from upstream, and clearly it isn't getting updates from upstream either. I saw closed tickets #20 #190 and #293. Ideally you'd use Go's standard crypto/ed25519 and so maybe there could be more work to see if you can optimize the code and submit issues/requests to the upstream project about any performance issues. This could also be highly dependent on the architecture and OS that the benchmarking was done on as well, so maybe more tests would be warranted in multiple environments.

Ideally, if there is a need to continue to use libsodium indefinitely you'd track your changes using a submodule and then create a patch file to continually test against upstream's latest release. Currently, we're just expected to trust your fork of libsodium which no on in their right mind would do.

## Scope/Requirements

- [ ] Simplify build to use standard Go tools.
- [ ] Remove excessive use of Makefile and provide instructions on building without.
- [ ] Cleanup docs and don't ask developers to blindly run an inception of scripts to setup environments. Running any script blindly is bad security practice.
- [ ] Update libsodium to upstream release and use submodule or patch for libsodium source during install.
- [ ] Revisit implementing Go's standard crypto/ed25519 package and submit track any issues upstream.
- [ ] Implement VRF in Go module to remove CGO requirement.
- [ ] Consider consolidating CLI tools into subcommands to reduce complexity.

## Urgency/Relative Priority
Would love to see this become a priority, as I think once it is done that it could be more easily adopted into Linux package repos and allow developers/engineers to get familiar with the tools a lot quicker and safely.

**Comment by tsachiherman** [comment]
@braderhart

All the issues you've mentioned above are on our boilerplate and being handled in one way or another. Some of the issues can't be easily solved, and might take a while before being addressed.

Note that none of the issues here would ever have a high priority since everything on the algorand toolchains works correctly ( from functional standpoint ). These build changes would primarily makes 3rd party integration *slightly* easier, which is nice to have, but not a blocker.

Regarding the VRF - it was implemented inside libsodium, so that it can take advantage of some of the construct available there, which would improve the performance. This is *required* in order to get to the advertised performance goals.
( nevertheless, I'm not saying that there aren't other ways to achieve the same goals )

The Go's standard crypto/ed25519, is a good implementation - but too slow for our purposes.

As for your comment regarding updating the libsodium upstream - there was a PR to the libsodium, which included our VRF implementation. That PR was rejected. At this time, no newer version of libsodium have any relevant changes, and therefore - we didn't update it. If you can point to a specific change in the libsodium that would benefit the Algorand platform, please open a concrete ticket.

**Comment by tmc** [comment]
@tsachiherman is there an issue/branch/etc. benchmarking the pure go implementation and the libsodium fork? I'm sure some crypto and performance interested folks would love to examine the delta and try to figure out how a pure go approach could be improved.

**Comment by brianolson** [comment]
iirc, I tested Go crypto/ed25519 vs libsodium and the speed difference was about 2:1. The Algorand Go SDK uses the pure go implementation to make that a cleaner build. I think a client might be expected to manage dozens or hundreds of ed25519 operations per second, but `algod` will be doing thousands per second and needs the fastest code. CPU profiling always turns up ed25519 operations as a substantial chunk of CPU time.

**Comment by braderhart** [comment]
I'm wondering what type of verification is being done from Go's implementation of ed25519? I did submit a PR that upgrades Algorand's code to use the latest upstream libsodium release. This at least helps somewhat in code maintenance and build verification, but was told that there was no reliable tests to make sure it was compatible with existing signature signing:

https://github.com/algorand/go-algorand/pull/2136

I would imagine the Go implementation would have the same concerns where the testing would need to be automated for changes in the crypto package. In terms of performance, 2:1 does seem like quite a bit, however it may be best to make this an option when building and also work with upstream Go to see if performance can be improved. Non-investor node operators really see no incentive to use the fastest code per say. If libsodium was the only CGO package required it wouldn't be that big of a deal for now, but there are at least a couple other CGO dependencies (`hidapi` & `libusb`) , and like Algorand they include their [source](https://github.com/karalabe/hid) copy of upstream projects.

I personally don't like the way the entire libsodium source is included in the repo, because it makes it more challenging to audit differences from upstream. If efforts could go into abstracting the libsodium changes and building automated tests to build from and track libsodium upstream, then researchers would only need to audit Algorand's changes rather than the entire libsodium fork.

Ideally at some point the build process would not rely on a bunch of shell scripts. It is really asking a lot of a user interested in building a Go project to run a large inception of shell scripts to build a project, when installing Go packages are meant to be done through a couple commands. At the very least, it would be ideal to have instructions for building without needing to rely on a bunch of shell scripts. That was really my biggest concerns, not so much relying on CGO.

**Comment by winder** [comment]
@braderhart I share your desire to simplify the shell scripts. There are a number of improvements that could be made to make the project more approachable for developers like yourself. That said, if you review these scripts I think you'll find that they're not as complicated as the number of scripts may suggest. There are some nice productivity features in them for things like automatically managing a local devnet node. The scripts also tie into some of our standard deployment tools, like `update.sh` which allow for continuous testing of the production deployment procedures.

If you prefer to skip the nonstandard helpers then `make buildsrc` may be what you're looking for. The actual compilation process starts there. It uses `go install` like you would expect.

The `libsodium` concerns are valid, but changes to this code must be taken very seriously. The code has already been audited so we cannot change it without good reason. As Tsachi mentioned we are working to improve things, if you would like to review some of that progress you can see it here: #2193.

And thanks for the feedback!

**Comment by abebeos** [comment]
@winder, others, can someone please recreate the original issue text (whilst leaving the last text intact)?

Would look somehow like this:

---

During the voyage of life, remember to keep an eye out for a fair wind; batten
down during a storm; hail all passing ships; and fly your colors proudly.

### ((Original Issue Text Follows))
---

### Summary

This all started when I just wanted to get started using the CLI tools, and the instructions on the website suggested that I run a script on my computer. After following the script, it was just too convoluted for me to reliably trust that it would even be a good idea. I have built multiple Go applications from source, including those that use C libraries.

When building go-algorand from source, it is currently expected that people run a Makefile that is c
[...]
Would love to see this become a priority, as I think once it is done that it could be more easily adopted into Linux package repos and allow developers/engineers to get familiar with the tools a lot quicker and safely.
